docs: distributed topology through replica clusters (#5133)
cloudnative-pg/cloudnative-pg: docs/src/replica_cluster.md

# Replica clusters

> A replica cluster is a separate CloudNativePG `Cluster` resource designed to
> replicate data from another PostgreSQL instance, ideally also managed by
> CloudNativePG.

Reviewer: ```suggestion
A replica cluster is a CloudNativePG `Cluster` resource designed to
replicate data from another PostgreSQL instance, ideally also managed by
CloudNativePG.
```

be a primary cluster or another replica cluster (cascading replication).

### Bootstrapping a Replica Cluster

The first step is to bootstrap the replica cluster using one of the following
methods:

- **Streaming replication** via `pg_basebackup`
- **Recovery from a volume snapshot**
- **Recovery from a Barman Cloud backup** in an object store

For detailed instructions on cloning a PostgreSQL server using `pg_basebackup`
(streaming) or recovery (volume snapshot or object store), refer to the
["Bootstrap" section](bootstrap.md#bootstrap-from-another-cluster).

### Configuring Replication

Once the base backup for the replica cluster is available, you need to define
how changes will be replicated from the origin using PostgreSQL continuous
recovery. There are three main options:

1. **Streaming Replication**: Set up streaming replication between the replica
   cluster and the source. This method requires configuring network connections
   and implementing appropriate administrative and security measures to ensure
   seamless data transfer.
2. **WAL Archive**: Utilize the WAL (Write-Ahead Logging) archive stored in an

>    object store. WAL files are regularly transferred from the source to the object

Reviewer: ```suggestion
   object store. WAL files are regularly transferred from the source cluster to the object
```

two Kubernetes clusters located in Southern Europe and Central Europe.

In this scenario, assume you have CloudNativePG installed in the Southern
Europe Kubernetes cluster, with a PostgreSQL `Cluster` named `cluster-eu-south`
acting as the primary. This cluster has continuous backup configured with a
local object store. This object store is also accessible by the PostgreSQL
`Cluster` named `cluster-eu-central`, installed in the Central European
Kubernetes cluster. Initially, `cluster-eu-central` functions as a replica
cluster. Following a symmetric approach, it also has a local object store for
continuous backup, which needs to be read by `cluster-eu-south`. The recovery
in this setup relies solely on WAL shipping, with no streaming connection
between the two clusters.

Here’s how you would configure the `externalClusters` section for both
`Cluster` resources:

```yaml
# Distributed topology configuration
externalClusters:
  - name: cluster-eu-south
    barmanObjectStore:
      destinationPath: s3://cluster-eu-central/
      # Additional configuration
  - name: cluster-eu-central
    barmanObjectStore:
      destinationPath: s3://cluster-eu-south/

>       # Additional configuration

Reviewer: My brain is twisted at this point.
The `south` cluster will use the `cluster-eu-central` as source, but the storage for `cluster-eu-central` is in the south.
So, we do a crossing here, and then we do another crossing in the replica section below.
But, the storage for the `south` cluster is likely in the south, no?

availability and disaster recovery through controlled switchover of a primary
PostgreSQL cluster using declarative configuration.

Controlled switchover in a distributed topology is a two-step process
involving:

- Demotion of a primary cluster to a replica cluster
- Promotion of a replica cluster to a primary cluster

These processes are described below.

### Demoting a Primary to a Replica Cluster

CloudNativePG provides the functionality to demote a primary cluster to a
replica cluster. This action is typically planned when transitioning the
primary role from one data center to another. The process involves demoting the
current primary cluster (e.g., `cluster-eu-south`) to a replica cluster and
subsequently promoting the designated replica cluster (e.g.,
`cluster-eu-central`) to primary when fully synchronized.

Provided you have defined an external cluster in the current primary `Cluster`
resource that points to the replica cluster that's been selected to become the
new primary, all you need to do is change the `primary` field as follows:

```yaml
replica:
  primary: cluster-eu-central
  source: cluster-eu-central
```

> When the primary PostgreSQL cluster is shut down, write operations are no
> longer possible. CloudNativePG then:

Reviewer: ```suggestion
When the primary PostgreSQL cluster is demoted, write operations are no
longer possible. CloudNativePG then:
```

    `promotionToken` for `cluster-eu-central`.

You can verify the role change using the `cnpg` plugin, checking the status of
the cluster:

```shell
kubectl cnpg status cluster-eu-south
```

### Promoting a Replica to a Primary Cluster

To promote a PostgreSQL replica cluster (e.g., `cluster-eu-central`) to a
primary cluster and make the designated primary an actual primary instance,
you need to perform the following steps simultaneously:

1. Set the `.spec.replica.primary` to the name of the current replica cluster
   to be promoted (e.g., `cluster-eu-central`).
2. Set the `.spec.replica.promotionToken` with the value obtained from the
   former primary cluster (refer to ["Demoting a Primary to a Replica Cluster"](replica_cluster.md#demoting-a-primary-to-a-replica-cluster)).

The updated `replica` section in `cluster-eu-central`'s spec should look like
this:

```yaml
replica:
  primary: cluster-eu-central
  promotionToken: <PROMOTION_TOKEN>
  source: cluster-eu-south
```

!!! Warning
    It is crucial to apply the changes to the `primary` and `promotionToken`
    fields simultaneously. If the promotion token is omitted, a failover will be

>     triggered, necessitating a rebuild of the former primary.

Reviewer: yikes. I hope we offer a plugin command to handle making this change safely

Reviewer 2: yes, we need to create a `kubectl cnpg` subcommand for this
